### PR TITLE
register ``structured_to_unstructured``

### PR DESCRIPTION
**Pull request by nstarman**
Signed-off-by: Nathaniel Starkman (@nstarman) <nstarkman@protonmail.com>

It works with any structured quantity that has just one unit, or compatible units.

### Checklist for package maintainer(s)

This checklist is meant to remind the package maintainer(s) who will review this pull request of some common things to look for. This list is not exhaustive.

- [ ] Do the proposed changes actually accomplish desired goals?
- [ ] Do the proposed changes follow the [Astropy coding guidelines](https://docs.astropy.org/en/latest/development/codeguide.html)?
- [ ] Are tests added/updated as required? If so, do they follow the [Astropy testing guidelines](https://docs.astropy.org/en/latest/development/testguide.html)?
- [ ] Are docs added/updated as required? If so, do they follow the [Astropy documentation guidelines](https://docs.astropy.org/en/latest/development/docguide.html#astropy-documentation-rules-and-guidelines)?
- [ ] Is rebase and/or squash necessary? If so, please provide the author with appropriate instructions. Also see ["When to rebase and squash commits"](https://docs.astropy.org/en/latest/development/when_to_rebase.html).
- [ ] Did the CI pass? If no, are the failures related? If you need to run daily and weekly cron jobs as part of the PR, please apply the `Extra CI` label.
- [ ] Is a change log needed? If yes, did the change log check pass? If no, add the `no-changelog-entry-needed` label. If this is a manual backport, use the `skip-changelog-checks` label unless special changelog handling is necessary.
- [ ] Is a milestone set? Milestone must be set but `astropy-bot` check might be missing; do not let the green checkmark fool you.
- [ ] At the time of adding the milestone, if the milestone set requires a backport to release branch(es), apply the appropriate `backport-X.Y.x` label(s) *before* merge.

### CODE REVIEW

**Comment by nstarman** [comment]
@mhvk, I'm not understanding the error that pops up. I thought ``@function_helper`` takes care of the registration...

File: astropy/units/quantity_helper/function_helpers.py [hunk 1]
```diff
@@ -1042,3 +1043,26 @@ def eig(a, *args, **kwargs):
...
+def structured_to_unstructured(arr, *args, **kwargs):
+    """
+    Convert a structured quantity to an unstructured one.
+    This only works if all the units are compatible.
+    """
+    units = arr.unit.values()
+    unit = units[0]
+
+    # all units are equal
+    if all(np.equal(units, units[0])):
```
Comment (mhvk) [inline astropy/units/quantity_helper/function_helpers.py:1058]: Just a fly-by comment for now: I think I'd write this more simply by creating a new structure unit which has `units[0]` in all its fields, and then doing `.to_value`, i.e., something along the lines of this trial:
```
In [15]: a = [(5, (4., 3.))] * u.Unit('m, (cm, um)')

In [16]: def replace_unit(x):
    ...:     if isinstance(x, u.StructuredUnit):
    ...:         return x._recursively_apply(replace_unit)
    ...:     else:
    ...:         return u.m
    ...: 

In [17]: a.unit._recursively_apply(replace_unit)
Out[17]: Unit("(m, (m, m))")

In [18]: a.to(a.unit._recursively_apply(replace_unit))
Out[18]: <Quantity [(5., (0.04, 3.e-06))] (m, (m, m))>

In [19]: a.to_value(a.unit._recursively_apply(replace_unit))
Out[19]: 
array([(5., (0.04, 3.e-06))],
      dtype=[('f0', '<f8'), ('f1', [('f0', '<f8'), ('f1', '<f8')])])

```

Comment (nstarman) [reply astropy/units/quantity_helper/function_helpers.py:1058]: Thanks! That's a much cleaner method to transform the unit.

**Comment by mhvk** [comment]
This looks good. The failure is because I check that coverage of the various numpy modules is complete, and here you add a function from a sub-module that is not covered at all (i.e., there now are more functions covered than the test code knows exist!) Obviously, would be nice to cover all of `recfunctions`, but fine to for now to just special-case it in the coverage tests.

File: astropy/units/tests/test_structured.py [hunk 2]
```diff
@@ -579,6 +582,39 @@ def test_zeros_ones_like(self, func):
...
+
+        # and if the dtype is nested
+        struct = [(5, (400.0, 3e6))] * u.Unit('m, (cm, um)')
+        unstruct = rfn.structured_to_unstructured(struct)
+        assert_array_equal(unstruct, [[5, 4, 3]] * u.m)
+
+    def test_unstructured_to_structured(self):
+        # can't structure something that's already structured
+        unstruct = [1, 2, 3] * u.m
+        dtype=np.dtype([("f1", int), ("f2", int), ("f3", int)])
```
Comment (mhvk) [inline astropy/units/tests/test_structured.py:608]: Absolute and total nitpick in this case, but I like not to encourage `int` for quantity as it generally makes no sense with units; can we just make it `float`?

File: docs/changes/units/12486.feature.rst [hunk 3]
```diff
@@ -0,0 +1 @@
+Let ``numpy.lib.structured_to_unstructured`` work with structured quantities.
```
Comment (mhvk) [inline docs/changes/units/12486.feature.rst:1]: Also supports `unstructured_to_structured` now!

File: astropy/units/quantity_helper/function_helpers.py [hunk 4]
```diff
@@ -118,7 +128,10 @@
     np.save, np.savez, np.savetxt, np.savez_compressed,
     # Polynomials
     np.poly, np.polyadd, np.polyder, np.polydiv, np.polyfit, np.polyint,
-    np.polymul, np.polysub, np.polyval, np.roots, np.vander}
+    np.polymul, np.polysub, np.polyval, np.roots, np.vander,
+    # record array functions
+    rfn.rec_append_fields, rfn.rec_drop_fields, rfn.rec_join,
```
Comment (mhvk) [inline astropy/units/quantity_helper/function_helpers.py:133]: Why are these added to `ignored` while the others are in `TOSUPPORT`?

File: astropy/units/quantity_helper/function_helpers.py [hunk 5]
```diff
@@ -107,6 +108,15 @@
 # Could be supported if we had a natural logarithm unit.
 UNSUPPORTED_FUNCTIONS |= {np.linalg.slogdet}
 
+# TODO! support whichever of these functions it makes sense to support
+TOSUPPORT_FUNCTIONS = {
```
Comment (mhvk) [inline astropy/units/quantity_helper/function_helpers.py:112]: maybe `TO_BE_DECIDED_FUNCTIONS` (or `TBD_FUNCTIONS`)?

File: astropy/units/tests/test_quantity_non_ufuncs.py [hunk 6]
```diff
@@ -2078,4 +2088,4 @@ def test_all_included(self):
     # untested_function is created using all_wrapped_functions
     @needs_array_function
     def test_ignored_are_untested(self):
-        assert IGNORED_FUNCTIONS == untested_functions
+        assert IGNORED_FUNCTIONS == (untested_functions - TOSUPPORT_FUNCTIONS)
```
Comment (mhvk) [inline astropy/units/tests/test_quantity_non_ufuncs.py:2091]: I'd write this as `assert IGNORED_FUNCTIONS | TOSUPPORT_FUNCTIONS == untested_functions`

File: astropy/units/tests/test_structured.py [hunk 7]
```diff
@@ -5,12 +5,15 @@
...
 import pytest
 import numpy as np
+import numpy.lib.recfunctions as rfn
 from numpy.testing import assert_array_equal
 
 from astropy import units as u
 from astropy.units import StructuredUnit, Unit, UnitBase, Quantity
 from astropy.utils.masked import Masked
 
+from .test_quantity_non_ufuncs import CoverageMeta
```
Comment (mhvk) [inline astropy/units/tests/test_structured.py:15]: One disadvantage of putting the tests here is that if one now tests just `pytest test_quantity_non_ufuncs.py`, one will get a failure. My sense would be to move the tests, especially since the actual tests do not really use the class setup much (you could still keep that test for failure here).

File: astropy/units/quantity_helper/function_helpers.py (hunk 4 above)
Comment (nstarman) [reply astropy/units/quantity_helper/function_helpers.py:133]: Because they work on record arrays, which are deprecated.
Does Quantity even work with a record array?

File: astropy/units/tests/test_structured.py [hunk 8]
```diff
@@ -579,6 +582,39 @@ def test_zeros_ones_like(self, func):
...
+        unstruct = rfn.structured_to_unstructured(struct)
+        assert_array_equal(unstruct, [0, 0, 600] * u.eV)
+
+        # and if the dtype is nested
+        struct = [(5, (400.0, 3e6))] * u.Unit('m, (cm, um)')
+        unstruct = rfn.structured_to_unstructured(struct)
+        assert_array_equal(unstruct, [[5, 4, 3]] * u.m)
+
+    def test_unstructured_to_structured(self):
+        # can't structure something that's already structured
```
[the remark this answers is not in the document]
Comment (nstarman) [reply astropy/units/tests/test_structured.py:606]: Yes,  that was mean to introduce ``with pytest.raises(ValueError, match=...`` below.

File: astropy/units/tests/test_structured.py (hunk 2 above)
Comment (nstarman) [reply astropy/units/tests/test_structured.py:608]: SGTM

File: astropy/units/quantity_helper/function_helpers.py [hunk 9]
```diff
@@ -1042,3 +1055,55 @@ def eig(a, *args, **kwargs):
...
+    unit : `astropy.units.Unit`
+
+    Returns
+    -------
+    `astropy.units.Unit` or tuple
+    """
+    if dtype.fields is None:
+        return unit
+
+    us = [_build_structured_unit(v[0], unit) for v in dtype.fields.values()]
```
[the remark this answers is not in the document]
Comment (nstarman) [reply astropy/units/quantity_helper/function_helpers.py]: Done.

File: astropy/units/tests/test_structured.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (nstarman) [reply astropy/units/tests/test_structured.py:594]: Moved.

File: astropy/units/tests/test_structured.py (hunk 2 above)
Comment (nstarman) [reply astropy/units/tests/test_structured.py]: Done.

File: docs/changes/units/12486.feature.rst (hunk 3 above)
Comment (nstarman) [reply docs/changes/units/12486.feature.rst]: ✅

File: astropy/units/quantity_helper/function_helpers.py (hunk 5 above)
Comment (nstarman) [reply astropy/units/quantity_helper/function_helpers.py]: ✅

File: astropy/units/tests/test_quantity_non_ufuncs.py (hunk 6 above)
Comment (nstarman) [reply astropy/units/tests/test_quantity_non_ufuncs.py]: ✅

File: astropy/units/tests/test_structured.py (hunk 7 above)
Comment (nstarman) [reply astropy/units/tests/test_structured.py:15]: ✅

File: astropy/units/quantity_helper/function_helpers.py (hunk 4 above)
Comment (mhvk) [reply astropy/units/quantity_helper/function_helpers.py:133]: OK, that makes sense, and I now see that I could have read that from the comment. Maybe make it a bit more explicit? E.g., "functions taking record arrays (which are deprecated)"

File: astropy/units/quantity_helper/function_helpers.py [hunk 10]
```diff
@@ -1042,3 +1055,54 @@ def eig(a, *args, **kwargs):
...
+        return unit
+
+    return tuple(_build_structured_unit(v[0], unit) for v in dtype.fields.values())
+
+
+@function_helper(module=np.lib.recfunctions)
+def unstructured_to_structured(arr, dtype, *args, **kwargs):
+    """
+    Convert a structured quantity to an unstructured one.
+    This only works if all the units are compatible.
```
Comment (mhvk) [inline astropy/units/quantity_helper/function_helpers.py:1102]: This works always, correct? Given that, I'd just remove the docstring, so that when I get around to making a nice sphinx page of the overrides (like https://docs.astropy.org/en/latest/utils/masked/index.html#module-astropy.utils.masked.function_helpers for `Masked), it will only list those that are not obvious.

File: astropy/units/quantity_helper/function_helpers.py [hunk 11]
```diff
@@ -1042,3 +1055,54 @@ def eig(a, *args, **kwargs):
     from astropy.units import dimensionless_unscaled
 
     return (a.value,)+args, kwargs, (a.unit, dimensionless_unscaled), None
+
+
+@function_helper(module=np.lib.recfunctions)
+def structured_to_unstructured(arr, *args, **kwargs):
+    """
+    Convert a structured quantity to an unstructured one.
+    This only works if all the units are compatible.
```
Comment (mhvk) [inline astropy/units/quantity_helper/function_helpers.py:1064]: Add a line break for later typesetting (see other comment)

File: astropy/units/tests/test_quantity_non_ufuncs.py [hunk 12]
```diff
@@ -2033,6 +2034,53 @@ def test_eigvalsh(self):
...
+        struct = u.Quantity((0, 0, 0.6), u.Unit("(eV, eV, eV)"))
+        unstruct = rfn.structured_to_unstructured(struct)
+        assert_array_equal(unstruct, [0, 0, 0.6] * u.eV)
+
+        # also if the units are convertible
+        struct = u.Quantity((0, 0, 0.6), u.Unit("(eV, eV, keV)"))
+        unstruct = rfn.structured_to_unstructured(struct)
+        assert_array_equal(unstruct, [0, 0, 600] * u.eV)
+
+        struct = u.Quantity((0, 0, 1.7827e-33), u.Unit("(eV, eV, g)"))
```
Comment (mhvk) [inline astropy/units/tests/test_quantity_non_ufuncs.py:2054]: Nice check on the equivalency!

File: astropy/units/tests/test_structured.py (hunk 7 above)
Comment (mhvk) [reply astropy/units/tests/test_structured.py:15]: Still to remove the import!

File: astropy/units/tests/test_structured.py [hunk 13]
```diff
@@ -558,7 +561,7 @@ def test_setitem(self):
         assert np.all(q_pv_t['pv'] == (1., 0.5) * self.pv_unit)
 
 
-class TestStructuredQuantityFunctions(StructuredTestBaseWithUnits):
+class TestStructuredQuantityFunctions(StructuredTestBaseWithUnits, metaclass=CoverageMeta):
```
Comment (mhvk) [inline astropy/units/tests/test_structured.py:564]: I think the metaclass is no longer needed.

**Review by mhvk — COMMENTED** [review]
Nitpicks only left

File: astropy/units/tests/test_structured.py (hunk 7 above)
Comment (nstarman) [reply astropy/units/tests/test_structured.py:15]: Ah. IMO that would be nice as pre-commit hook : to check for unused imports and error unless #noqa

File: astropy/units/quantity_helper/function_helpers.py (hunk 10 above)
Comment (nstarman) [reply astropy/units/quantity_helper/function_helpers.py:1102]: Yes, I believe it should always work. Removing now.

File: astropy/units/tests/test_quantity_non_ufuncs.py (hunk 12 above)
Comment (nstarman) [reply astropy/units/tests/test_quantity_non_ufuncs.py:2054]: It was nice to realize it worked. I hadn't intentionally ensured it would.

File: astropy/units/quantity_helper/function_helpers.py (hunk 11 above)
Comment (nstarman) [reply astropy/units/quantity_helper/function_helpers.py:1064]: ✅

File: astropy/units/tests/test_structured.py (hunk 7 above)
Comment (mhvk) [reply astropy/units/tests/test_structured.py]: Yes, along with other PEP stuff...

File: astropy/units/quantity_helper/function_helpers.py (hunk 10 above)
Comment (nstarman) [reply astropy/units/quantity_helper/function_helpers.py]: ✅

File: astropy/units/tests/test_structured.py (hunk 13 above)
Comment (nstarman) [reply astropy/units/tests/test_structured.py]: Agreed. Removing.

Comment (nstarman) [reply astropy/units/tests/test_structured.py]: ✅

**Review by mhvk — APPROVED** [review]
Great, let's get this in!  Thanks, @nstarman!

**Comment by mhvk** [comment]
Well, let's get it in once tests passed - seems like things are a bit stuck right now...

**Comment by pllim** [comment]
I have updated the CI today. You need to rebase for the CI "required" stuff to be normal again. I apologize for any inconvenience caused.

**Comment by mhvk** [comment]
@pllim - I guess it just ends up we reviewers have to be more careful, especially just before the feature freeze... Thanks for trying to help solve this, even if it turns out this particular cure was worse than the problem!
